File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/calcudoku/domain/CalcudokuGame.java
package com.zwstudio.logicpuzzlesandroid.puzzles.calcudoku.domain;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame;
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.Graph;
import com.zwstudio.logicpuzzlesandroid.common.domain.GridDots;
import com.zwstudio.logicpuzzlesandroid.common.domain.GridLineObject;
import com.zwstudio.logicpuzzlesandroid.common.domain.Node;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fj.F2;

import static fj.data.HashMap.fromMap;
import static fj.data.List.iterableList;

/**
 * Created by zwvista on 2016/09/29.
 */

public class CalcudokuGame extends CellsGame<CalcudokuGame, CalcudokuGameMove, CalcudokuGameState> {
    public static Position offset[] = {
            new Position(-1, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(0, -1),
    };
    public static Position offset2[] = {
            new Position(0, 0),
            new Position(1, 1),
            new Position(1, 1),
            new Position(0, 0),
    };
    public static int dirs[] = {1, 0, 3, 2};

    public List<List<Position>> areas = new ArrayList<>();
    public Map<Position, Integer> pos2area = new HashMap<>();
    public GridDots dots;

    public char[] objArray;
    public Map<Position, CalcudokuHint> pos2hint = new HashMap<>();

    public char get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public char get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, char obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, char obj) {
        set(p.row, p.col, obj);
    }

    public CalcudokuGame(List<String> layout, GameInterface<CalcudokuGame, CalcudokuGameMove, CalcudokuGameState> gi, GameDocumentInterface gdi) {
        super(gi, gdi);
        size = new Position(layout.size(), layout.get(0).length() / 4);
        dots = new GridDots(rows() + 1, cols() + 1);
        objArray = new char[rows() * cols()];
        Graph g = new Graph();
        Map<Position, Node> pos2node = new HashMap<>();
        for (int r = 0; r < rows(); r++) {
            String str = layout.get(r);
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                char ch1 = str.charAt(c * 4);
                String s = str.substring(c * 4 + 1, c * 4 + 3);
                char ch2 = str.charAt(c * 4 + 3);
                set(p, ch1);
                Node node = new Node(p.toString());
                g.addNode(node);
                pos2node.put(p, node);
                if (s.equals("  ")) continue;
                pos2hint.put(p, new CalcudokuHint() {{
                    op = ch2;
                    result = s.equals("  ") ? 0 : Integer.parseInt(s.trim());
                }});
            }
        }
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                char ch = get(p);
                if (ch == ' ') continue;
                for (Position os : CalcudokuGame.offset) {
                    Position p2 = p.add(os);
                    if (isValid(p2) && get(p2) == ch)
                        g.connectNode(pos2node.get(p), pos2node.get(p2));
                }
            }
        while (!pos2node.isEmpty()) {
            g.setRootNode(iterableList(pos2node.values()).head());
            List<Node> nodeList = g.bfs();
            List<Position> area = fromMap(pos2node).toStream().filter(e -> nodeList.contains(e._2())).map(e -> e._1()).toJavaList();
            int n = areas.size();
            char ch = get(area.get(0));
            for (Position p : area) {
                pos2area.put(p, n);
                pos2node.remove(p);
                for (int i = 0; i < 4; i++) {
                    Position p2 = p.add(CalcudokuGame.offset[i]);
                    char ch2 = !isValid(p2) ? '.' : get(p2);
                    if (ch2 != ch)
                        dots.set(p.add(CalcudokuGame.offset2[i]), CalcudokuGame.dirs[i], GridLineObject.Line);
                }
            }
            areas.add(area);
        }
        for (int r = 0; r < rows(); r++) {
            dots.set(r, 0, 2, GridLineObject.Line);
            dots.set(r + 1, 0, 0, GridLineObject.Line);
            dots.set(r, cols(), 2, GridLineObject.Line);
            dots.set(r + 1, cols(), 0, GridLineObject.Line);
        }
        for (int c = 0; c < cols(); c++) {
            dots.set(0, c, 1, GridLineObject.Line);
            dots.set(0, c + 1, 3, GridLineObject.Line);
            dots.set(rows(), c, 1, GridLineObject.Line);
            dots.set(rows(), c + 1, 3, GridLineObject.Line);
        }

        CalcudokuGameState state = new CalcudokuGameState(this);
        states.add(state);
        levelInitilized(state);
    }

    private boolean changeObject(CalcudokuGameMove move, F2<CalcudokuGameState, CalcudokuGameMove, Boolean> f) {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size()).clear();
            moves.subList(stateIndex, states.size()).clear();
        }
        CalcudokuGameState state = cloner.deepClone(state());
        boolean changed = f.f(state, move);
        if (changed) {
            states.add(state);
            stateIndex++;
            moves.add(move);
            moveAdded(move);
            levelUpdated(states.get(stateIndex - 1), state);
        }
        return changed;
   }

    public boolean switchObject(CalcudokuGameMove move) {
        return changeObject(move, (state, move2) -> state.switchObject(move2));
    }

    public boolean setObject(CalcudokuGameMove move) {
        return changeObject(move, (state, move2) -> state.setObject(move2));
    }

    public int getObject(Position p) {
        return state().get(p);
    }

    public int getObject(int row, int col) {
        return state().get(row, col);
    }

    public HintState getRowState(int row) {
        return state().row2state[row];
    }

    public HintState getColState(int col) {
        return state().col2state[col];
    }

    public HintState getPosState(Position p) {
        return state().pos2state.get(p);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/calcudoku/domain/CalcudokuGame.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.calcudoku.domain

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.*
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState
import fj.F2
import fj.P2
import java.util.*

class CalcudokuGame(layout: List<String>, gi: GameInterface<CalcudokuGame, CalcudokuGameMove, CalcudokuGameState>, gdi: GameDocumentInterface) : CellsGame<CalcudokuGame, CalcudokuGameMove, CalcudokuGameState>(gi, gdi) {
    var areas: MutableList<List<Position>> = ArrayList()
    var pos2area: MutableMap<Position?, Int?> = HashMap()
    var dots: GridDots
    var objArray: CharArray
    var pos2hint: MutableMap<Position, CalcudokuHint> = HashMap()
    operator fun get(row: Int, col: Int): Char {
        return objArray[row * cols() + col]
    }

    operator fun get(p: Position): Char {
        return get(p.row, p.col)
    }

    operator fun set(row: Int, col: Int, obj: Char) {
        objArray[row * cols() + col] = obj
    }

    operator fun set(p: Position, obj: Char) {
        set(p.row, p.col, obj)
    }

    private fun changeObject(move: CalcudokuGameMove?, f: F2<CalcudokuGameState?, CalcudokuGameMove?, Boolean>): Boolean {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size).clear()
            moves.subList(stateIndex, states.size).clear()
        }
        val state = cloner.deepClone(state())
        val changed = f.f(state, move)
        if (changed) {
            states.add(state)
            stateIndex++
            moves.add(move)
            moveAdded(move)
            levelUpdated(states[stateIndex - 1], state)
        }
        return changed
    }

    fun switchObject(move: CalcudokuGameMove?): Boolean {
        return changeObject(move, F2 { obj: CalcudokuGameState?, move: CalcudokuGameMove? -> obj!!.switchObject(move) })
    }

    fun setObject(move: CalcudokuGameMove?): Boolean {
        return changeObject(move, F2 { obj: CalcudokuGameState?, move: CalcudokuGameMove? -> obj!!.setObject(move) })
    }

    fun getObject(p: Position?): Int {
        return state()!![p]
    }

    fun getObject(row: Int, col: Int): Int {
        return state()!![row, col]
    }

    fun getRowState(row: Int): HintState? {
        return state()!!.row2state[row]
    }

    fun getColState(col: Int): HintState? {
        return state()!!.col2state[col]
    }

    fun getPosState(p: Position?): HintState? {
        return state()!!.pos2state[p]
    }

    companion object {
        var offset = arrayOf(
            Position(-1, 0),
            Position(0, 1),
            Position(1, 0),
            Position(0, -1))
        var offset2 = arrayOf(
            Position(0, 0),
            Position(1, 1),
            Position(1, 1),
            Position(0, 0))
        var dirs = intArrayOf(1, 0, 3, 2)
    }

    init {
        size = Position(layout.size, layout[0].length / 4)
        dots = GridDots(rows() + 1, cols() + 1)
        objArray = CharArray(rows() * cols())
        val g = Graph()
        val pos2node: MutableMap<Position, Node> = HashMap()
        for (r in 0 until rows()) {
            val str = layout[r]
            for (c in 0 until cols()) {
                val p = Position(r, c)
                val ch1 = str[c * 4]
                val s = str.substring(c * 4 + 1, c * 4 + 3)
                val ch2 = str[c * 4 + 3]
                set(p, ch1)
                val node = Node(p.toString())
                g.addNode(node)
                pos2node[p] = node
                if (s == "  ") continue
                pos2hint[p] = object : CalcudokuHint() {
                    init {
                        op = ch2
                        result = if (s == "  ") 0 else s.trim { it <= ' ' }.toInt()
                    }
                }
            }
        }
        for (r in 0 until rows()) for (c in 0 until cols()) {
            val p = Position(r, c)
            val ch = get(p)
            if (ch == ' ') continue
            for (os in offset) {
                val p2 = p.add(os)
                if (isValid(p2) && get(p2) == ch) g.connectNode(pos2node[p], pos2node[p2])
            }
        }
        while (!pos2node.isEmpty()) {
            g.setRootNode(fj.data.List.iterableList(pos2node.values).head())
            val nodeList = g.bfs()
            val area = fj.data.HashMap.fromMap(pos2node).toStream().filter { e: P2<Position, Node> -> nodeList.contains(e._2()) }.map { e: P2<Position, Node> -> e._1() }.toJavaList()
            val n = areas.size
            val ch = get(area[0])
            for (p in area) {
                pos2area[p] = n
                pos2node.remove(p)
                for (i in 0..3) {
                    val p2 = p.add(offset[i])
                    val ch2 = if (!isValid(p2)) '.' else get(p2)
                    if (ch2 != ch) dots[p.add(offset2[i]), dirs[i]] = GridLineObject.Line
                }
            }
            areas.add(area)
        }
        for (r in 0 until rows()) {
            dots[r, 0, 2] = GridLineObject.Line
            dots[r + 1, 0, 0] = GridLineObject.Line
            dots[r, cols(), 2] = GridLineObject.Line
            dots[r + 1, cols(), 0] = GridLineObject.Line
        }
        for (c in 0 until cols()) {
            dots[0, c, 1] = GridLineObject.Line
            dots[0, c + 1, 3] = GridLineObject.Line
            dots[rows(), c, 1] = GridLineObject.Line
            dots[rows(), c + 1, 3] = GridLineObject.Line
        }
        val state = CalcudokuGameState(this)
        states.add(state)
        levelInitilized(state)
    }
}